### PR TITLE
trace: Improve documentation

### LINKED ISSUE #907: tokio-trace: how to use Spans to represent concurrency?
I've been reading the code for tokio-trace, and I've got some questions about `Span`s and how they're intended to be used.

It looks like they're pretty good at representing strictly nested time-periods, as you'd get with a function call tree. But its unclear to me if the direct children of a parent `Span` can overlap with each other (ie, representing long-term concurrent operations).

For example, if I have a Future which is managing 10 concurrent connections which it initiates at once and then waits for each of them to produce a result, could I represent that as a parent `Span` with 10 sub-`Span` representing each connection (which in turn may have sub-spans, etc)? It seems like that would't work because there's always a notion of "the" current `Span`.

Or is this not the intended level of granularity? Should I be using `Span` just to track specific calls to `Future`/`Stream::poll` and derive tracing at that level? I can see how that would be possible, but it seems like it would be hard to reconstruct application-level states from that level of granularity.

The example doesn't really help clarify this, because it omits all actual async behaviour.

On a more general level, is there a way to represent a dependency relationship in the trace? For example, if task A is waiting for something from task B, can that be recorded?

### LINKED ISSUE #985: trace-core: Augment `Subscriber::record_follows_from` docs
Currently, reading the [`Subscriber::record_follows_from`](https://docs.rs/tokio-trace-core/0.1.0/tokio_trace_core/subscriber/trait.Subscriber.html#tymethod.record_follows_from) are extensive but lacks information of when this might be called. There exists the api in `tokio-trace` it self for a span to follow from another but this does not exist in `tokio-trace-core` and this to me is an area that could generate some confusion.

### LINKED ISSUE #1063: trace [docs]: explain field vs attribute/meta syntax
The docs don't currently explain the difference between `x: y` and `x = y` syntax in `span!` and `event!` macros. It would be useful if that existed, and also told you what possible `x`s you can set in `x:` context for both span and event.

### LINKED ISSUE #1064: trace [docs]: docs should make clear what a "target" is
Maybe I missed it, but it wasn't clear to me what a "target" was, or what I might set it to. @hawkw pointed me to the `log` docs, which say:

> A log request consists of a target, a level, and a body. A target is a string which defaults to the module path of the location of the log request, though that default may be overridden. Logger implementations typically use the target to filter requests based on some user configuration.

Having something similar, maybe with some more context, would be helpful.

### LINKED ISSUE #1065: trace [docs]: show examples of setting explicit parents
The span docs don't currently give examples of setting the `parent` attribute explicitly, nor do they talk about why you may want to do that.

@hawkw also mentioned that it might be good to support `parent` on `event!`.

### PR DESCRIPTION
**Pull request by hawkw**
## Motivation

Currently, some parts of the `tokio-trace` documentation is unclear, and
some concepts are not described adequately. In addition, the
organization of the docs can be improved and made less redundant. For
example, the lib-level documentation has examples of the shorthand
syntax for fields that is repeated in the span macro's documentation.
This could be changed so that the lib and module-level documentation
provides an _overview_ and the individual items go into greater detail.

## Solution

This branch reorganizes the documentation and adds better explanations
of some concepts that were previously unclear. In particular, I've
added:
 - examples of setting metadata and overriding span parents in the 
   macros
 - an explanation of what targets are,
 - added an explanation of the span follows-from relationship, and how
   it differs from parent/child relationships

I've also fixed up some typos and broken links.

Fixes #1063
Fixes #1064
Fixes #1065
Fixes #985
Fixes #907

### CODE REVIEW

**Comment by hawkw** [comment]
@jonhoo and @LucioFranco, since a lot of the additions here are from improvements you suggested, I'd love to get your feedback on this when you have the chance.

File: tokio-trace/src/span.rs [hunk 1]
```diff
@@ -64,10 +200,9 @@
 //!     type Error = ();
 //!
 //!     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-//!         self.span.in_scope(|| {
-//!             // Do actual future work
+//!         let _enter = self.span.enter();
```
Comment (davidbarsky) [inline tokio-trace/src/span.rs]: If this is a Mutex-like scopeguard, why assign to `_enter`?

Comment (davidbarsky) [reply tokio-trace/src/span.rs]: For some reason, I had the wrong understanding of how MutexGuards worked. Feel free to disregard.

File: tokio-trace/src/lib.rs [hunk 2]
```diff
@@ -27,9 +27,9 @@
 //! The core of `tokio-trace`'s API is composed of `Event`s, `Span`s, and
 //! `Subscriber`s. We'll cover these in turn.
```
Comment (jonhoo) [inline tokio-trace/src/lib.rs]: nit: maybe make the order here the same as the order they're explained in below?

File: tokio-trace/src/lib.rs [hunk 3]
```diff
@@ -49,153 +49,7 @@
 //! # }
```
Comment (jonhoo) [inline tokio-trace/src/lib.rs]: I think if we're not going to give any more stuff here on spans, it'd be good to make this example a little more expansive. I would also avoid using the word "context" here, because as far as I've gathered, it's entirely up to the subscriber whether or not work done below `.enter()` is considered "in the context of" the span. The span _only_ serves to build a pair of start and end events, and does not say anything about events "contained" within that.

File: tokio-trace/src/lib.rs [hunk 4]
```diff
@@ -49,153 +49,7 @@
...
-//! ```
-//!
-//! Depending on the circumstances, we might want to do either, or both. For
-//! example, if we want to know how long was spent in the loop overall, we would
-//! create a single span around the entire loop; whereas if we wanted to know how
-//! much time was spent in each individual iteration, we would enter a new span
-//! on every iteration.
+//! The [`span` module]'s documentation provides further details on how to use spans.
 //!
 //! ## Events
```
Comment (jonhoo) [inline tokio-trace/src/lib.rs]: This section reinforces the belief that a span establishes a "context", which again does not match my understand of what Spans actually mean?

Comment (jonhoo) [reply tokio-trace/src/lib.rs]: Further down, this text also says:
> Events are represented as a special case of spans — they are created, they may have fields added, and then they close immediately, without being entered.

Is that true? I thought events _weren't_ just "immediate spans"?

**Comment by jonhoo** [comment]
Reading through it now!

As a note for a possibly later PR, I think the "In libraries" is one of the most important and difficult sections, and it is currently basically empty.

File: tokio-trace/src/macros.rs [hunk 5]
```diff
@@ -2,19 +2,23 @@
...
 /// # Examples
 ///
-/// Creating a new span with no fields:
+/// Creating a new span:
 /// ```
-/// # #[macro_use]
-/// # extern crate tokio_trace;
-/// # use tokio_trace::Level;
+/// #[macro_use]
+/// extern crate tokio_trace;
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: Should this use 2018 syntax?

File: tokio-trace/src/macros.rs [hunk 6]
```diff
@@ -25,31 +29,77 @@
...
+/// # use tokio_trace::Level;
+/// # fn main() {
+/// let foo = span!(Level::INFO, "foo");
+//
+/// let bar = span!(Level::INFO, target: "bar_events", parent: &foo, "bar");
+/// # }
+/// ```
+///
+/// By default, the module path to the current Rust module will be used
+/// as the target, and the parent will be [determined contextually].
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: I don't think that what this links to is useful. Is "contextual spans" explained anywhere else (like in the `Span` docs)?

File: tokio-trace/src/macros.rs [hunk 7]
```diff
@@ -132,11 +182,13 @@
 /// };
 ///
 /// // `my_struct` will be recorded using its `fmt::Debug` implementation.
-/// let my_span = span!(Level::TRACE, "my span", foo = ?my_struct);
+/// span!(Level::TRACE, "my span", foo = ?my_struct);
+/// // This is equivalent to
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: Code above uses:
```rust
// is equivalent to:
```

File: tokio-trace/src/macros.rs [hunk 8]
```diff
@@ -151,7 +203,9 @@
 /// #     field: "Hello world!"
 /// # };
 /// // `my_struct.field` will be recorded using its `fmt::Display` implementation.
-/// let my_span = span!(Level::TRACE, "my span", foo = %my_struct.field);
+/// span!(Level::TRACE, "my span", foo = %my_struct.field);
+/// // This is equivalent to
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: Code above uses:
```rust
// is equivalent to:
```

File: tokio-trace/src/macros.rs [hunk 9]
```diff
@@ -151,7 +203,9 @@
...
+/// span!(Level::TRACE, "my span", foo = tokio_trace::field::display(&my_struct.field));
 /// # }
 /// ```
 ///
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: Should the line below us `%` and `?` instead of `display` and `debug`?

File: tokio-trace/src/macros.rs [hunk 10]
```diff
@@ -25,31 +29,77 @@
...
+/// let foo = span!(Level::INFO, "foo");
+//
+/// let bar = span!(Level::INFO, target: "bar_events", parent: &foo, "bar");
+/// # }
+/// ```
+///
+/// By default, the module path to the current Rust module will be used
+/// as the target, and the parent will be [determined contextually].
+///
+/// ## Recording Fields
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: I think it makes more sense for this section to come first, since you will set fields more frequently than you set attributes.

File: tokio-trace/src/macros.rs [hunk 11]
```diff
@@ -300,6 +360,13 @@ macro_rules! span {
 
 /// Constructs a span at the trace level.
 ///
+/// [Attributes] and [fields] are set using the same syntax as the [`span!`]
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: nit: I think I'd list fields first, since they are more common.

File: tokio-trace/src/macros.rs [hunk 12]
```diff
@@ -361,6 +428,13 @@ macro_rules! trace_span {
 
 /// Constructs a span at the debug level.
 ///
+/// [Attributes] and [fields] are set using the same syntax as the [`span!`]
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: nit: I think I'd list fields first, since they are more common.

File: tokio-trace/src/macros.rs [hunk 13]
```diff
@@ -422,6 +496,13 @@ macro_rules! debug_span {
 
 /// Constructs a span at the info level.
 ///
+/// [Attributes] and [fields] are set using the same syntax as the [`span!`]
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: nit: I think I'd list fields first, since they are more common.

File: tokio-trace/src/macros.rs [hunk 14]
```diff
@@ -483,6 +564,13 @@ macro_rules! info_span {
 
 /// Constructs a span at the warn level.
 ///
+/// [Attributes] and [fields] are set using the same syntax as the [`span!`]
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: nit: I think I'd list fields first, since they are more common.

File: tokio-trace/src/macros.rs [hunk 15]
```diff
@@ -543,6 +631,13 @@ macro_rules! warn_span {
 }
 /// Constructs a span at the error level.
 ///
+/// [Attributes] and [fields] are set using the same syntax as the [`span!`]
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: nit: I think I'd list fields first, since they are more common.

File: tokio-trace/src/macros.rs [hunk 16]
```diff
@@ -300,6 +360,13 @@ macro_rules! span {
 
 /// Constructs a span at the trace level.
 ///
+/// [Attributes] and [fields] are set using the same syntax as the [`span!`]
+/// macro.
+///
```
Comment (jonhoo) [inline tokio-trace/src/macros.rs]: I think a code block for each of these with:
```rust
let span = trace_span!("my span");
// is equivalent to:
let span = span!(Level::TRACE, "my span");
```

File: tokio-trace/src/span.rs [hunk 17]
```diff
@@ -1,6 +1,50 @@
-//! Spans represent periods of time in the execution of a program.
+//! Spans represent periods of time in which a program was executing in a
+//! particular context.
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: I thought a span explicitly did _not_ establish context, and was _only_ a period of time? Anything else is up to the subscriber, no?

File: tokio-trace/src/span.rs [hunk 18]
```diff
@@ -1,6 +1,50 @@
...
 //!
-//! # Entering a Span
+//! A span consists of [fields], user-defined key-value pairs of arbitrary data
+//! that describe the context the span represents, and [metadata], a fixed set
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: Other docs refer to these just as attributes. Should we use "metadata" everywhere, or "attributes" everywhere?

File: tokio-trace/src/span.rs [hunk 19]
```diff
@@ -1,6 +1,50 @@
...
+//! use tokio_trace::Level;
+//!
+//! # fn main() {
+//! /// Construct a new span at the `INFO` level named "my_span", with a single
+//! /// field named answer , with the value `42`.
+//! let my_span = span!(Level::INFO, "my_span", answer = 42);
+//! ```
+//!
+//! The documentation for the [`span!`] macro provides a complete set of
+//! examples for the various options that exist when creating spans.
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: "a complete set of" doesn't sound right. Very few things are complete.

File: tokio-trace/src/span.rs [hunk 20]
```diff
@@ -1,6 +1,50 @@
...
+//!
+//! The [`trace_span!`], [`debug_span!`], [`info_span!`], [`warn_span!`], and
+//! [`error_span!`] exist as shorthand for constructing spans at various
+//! verbosity levels.
+//!
+//! ## Recording Span Creation
+//!
+//! The [`Attributes`] type contains data associated with a span, and is
+//! provided to the [`Subscriber`] when a new span is created. It contains
+//! the span's metadata, the ID of the span's parent if one was explicitly set,
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: I think somewhere needs to talk about the relationship between spans and how it is or is not determined. When reading this sentence, I had several questions:

 - What if it isn't set?
 - What happens if I set it to something "wrong"?
 - What are the implications of setting one span as another's "parent"?

File: tokio-trace/src/span.rs [hunk 21]
```diff
@@ -1,6 +1,50 @@
...
+//! and any fields whose values were recorded when the span was constructed.
+//! The subscriber may then choose to cache the data for future use, record
+//! it in some manner, or discard it completely.
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: This last sentence is pretty vague given that I (currently) have no idea what a subscriber is if I have followed the docs thus far. I think elaborating a bit here would be worthwhile.

File: tokio-trace/src/span.rs [hunk 22]
```diff
@@ -1,6 +1,50 @@
...
+//!
+//! # The Span Lifecycle
+//!
+//! ## Entering a Span
 //!
 //! A thread of execution is said to _enter_ a span when it begins executing,
 //! and _exit_ the span when it switches to another context. Spans may be
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: What does "switches to another context" mean?

File: tokio-trace/src/span.rs [hunk 23]
```diff
@@ -1,6 +1,50 @@
...
+//! The [`Attributes`] type contains data associated with a span, and is
+//! provided to the [`Subscriber`] when a new span is created. It contains
+//! the span's metadata, the ID of the span's parent if one was explicitly set,
+//! and any fields whose values were recorded when the span was constructed.
+//! The subscriber may then choose to cache the data for future use, record
+//! it in some manner, or discard it completely.
+//!
+//! # The Span Lifecycle
+//!
+//! ## Entering a Span
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: I think this is missing an explanation of what "entering a span" even means in the first place.

File: tokio-trace/src/span.rs [hunk 24]
```diff
@@ -44,8 +90,98 @@
...
+//! **Note:** Since entering a span takes `&self`, and `Span`s are `Clone`,
+//! `Send`, and `Sync`, it is entirely valid for multiple threads to enter the
+//! same span concurrently.
 //!
-//! # The Span Lifecycle
+//! ## Span Relationships
+//!
+//! Spans form a tree structure — unless it is a root span, all spans have a
+//! _parent_, and may have one or more _children_. When a new span is created,
+//! the current span becomes the new span's parent. The total execution time of
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: This answers some of the questions from above, but raises a few more:

 - How is the "current" span determined?
 - What if I have entered multiple spans?
 - Is this actually guaranteed, or dependent on the subscriber?

File: tokio-trace/src/span.rs [hunk 25]
```diff
@@ -44,8 +90,98 @@
...
+//! a span consists of the time spent in that span and in the entire subtree
+//! represented by its children. Thus, a parent span always lasts for at least
+//! as long as the longest-executing span in its subtree.
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: Is that true? That seems weird. What if I have the following:
```rust
let a = trace_span!("a");
let b = trace_span!("b");
let ag = a.enter();
let bg = b.enter();
drop(ag);
// what happens to events here? are they in a or not? is this behavior guaranteed?
drop(bg);
```

File: tokio-trace/src/span.rs [hunk 26]
```diff
@@ -44,8 +90,98 @@
...
+//! // Although we have currently entered "bar", "baz"'s parent span
+//! // will be "foo".
+//! let baz = span!(Level::INFO, parent: &foo, "baz");
+//! # }
+//! ```
+//!
+//! A child span should typically be considered _part_ of its parent. For
+//! example, if a subscriber is recording the length of time spent in various
+//! spans, it should generally include the time spent in a span's children as
+//! part of that span's duration.
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: Does that mean that time in children is double-counted? Once in the parent and once in the child?

File: tokio-trace/src/span.rs [hunk 27]
```diff
@@ -44,8 +90,98 @@
...
+//! # }
+//! ```
+//!
+//! As an example, consider a listener task in a server. As the listener accepts
+//! incoming connections, it spawns new tasks that handle those connections. We
+//! might want to have a span representing the listener, and instrument each
+//! spawned handler task with its own span. We would want our instrumentation to
+//! record that the handler tasks were spawned as a result of the listener task.
+//! However, we might  not consider the handler tasks to be _part_ of the time
+//! spent in the listener taskm so we would not consider those spans children of
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]:
```suggestion
//! spent in the listener task, so we would not consider those spans children of
```

File: tokio-trace/src/span.rs [hunk 28]
```diff
@@ -117,23 +252,69 @@
 //! given ID, then no more handles to the span with that ID exist. The
```
Comment (jonhoo) [inline tokio-trace/src/span.rs]: There's a typo above:

> If there is only a single handle with the capacity to exit a span, dropping that handle **"close"** the span, since the capacity to enter it no longer exists.

Should say "closes"

File: tokio-trace/src/span.rs (hunk 18 above)
Comment (jonhoo) [reply tokio-trace/src/span.rs]: Actually, aren't the two distinct? Is there a difference between `Attributes` and `Metadata`? What is it? Which do you mean here?

File: tokio-trace/tokio-trace-core/src/metadata.rs [hunk 29]
```diff
@@ -7,8 +7,21 @@ use std::fmt;
 
 /// Metadata describing a [span] or [event].
```
Comment (jonhoo) [inline tokio-trace/tokio-trace-core/src/metadata.rs]: The `[event]` link here is broken.

**Comment by jonhoo** [comment]
The link to "metadata" from `Span::metadata` is broken (the one from `Event::metadata` is not).

File: tokio-trace/tokio-trace-core/src/metadata.rs (hunk 29 above)
Comment (jonhoo) [reply tokio-trace/tokio-trace-core/src/metadata.rs]: Sorry, it is broken in `tokio-trace::Metadata`, not in `tokio-trace-core::Metadata`

Comment (jonhoo) [reply tokio-trace/tokio-trace-core/src/metadata.rs]: Related to that, is there any way for the two to be the same? Why are they distinct types with different docs (the docs for `tokio-trace::Metadata` haven't been updated to reflect the change below for example.

File: tokio-trace/src/span.rs (hunk 18 above)
Comment (jonhoo) [reply tokio-trace/src/span.rs]: Is `Metadata` attributes _and_ fields maybe?

**Comment by jonhoo** [comment]
Changes look good overall! I left some notes :)

**Comment by LucioFranco** [comment]
I took a quick look through it looks great! Unfortunetly wont have time to review this but thank you for all the work @hawkw!

**Comment by hawkw** [comment]
@jonhoo thanks so much for the extremely thorough review, I really appreciate it!

File: tokio-trace/src/lib.rs (hunk 3 above)
Comment (hawkw) [reply tokio-trace/src/lib.rs]: > I think if we're not going to give any more stuff here on spans, it'd be good to make this example a little more expansive

I thought the docs would be more readable if more of the content in the lib-level docs was pulled out into module-level docs, so the reader isn't immediately presented with a massive wall of text. The hope was that the lib docs would serve as an overview/table of contents, and most of the actual documentation would live in the appropriate module. I'd love to know if you think this is actually a good way to structure things, or if it's not working for you?

That said, I can definitely add a more complete example here. I wasn't sure about including an example that shows a lot of concepts that haven't been introduced yet in the documentation, though?

> I would also avoid using the word "context" here, because as far as I've gathered, it's entirely up to the subscriber whether or not work done below .enter() is considered "in the context of" the span. The span only serves to build a pair of start and end events, and does not say anything about events "contained" within that.

I think there is an expectation that there's some _concept_ of "context", but the policy of what a "context" means is up to the subscriber. I'll try and edit these docs to make it clearer that this is provided by the subscriber implementation.

File: tokio-trace/src/lib.rs (hunk 4 above)
Comment (hawkw) [reply tokio-trace/src/lib.rs]: > Is that true? I thought events weren't just "immediate spans"?

That sentence is no longer correct, I think it was written when this _was_ the case and then it was never removed. I must've missed that while working on this branch, since I was mostly just looking at the span docs. Thanks for catching that!

File: tokio-trace/src/macros.rs (hunk 5 above)
Comment (hawkw) [reply tokio-trace/src/macros.rs]: We still have a CI build against Rust 1.26.0, so I don't think we can use the 2018 edition yet?

File: tokio-trace/tokio-trace-core/src/metadata.rs (hunk 29 above)
Comment (hawkw) [reply tokio-trace/tokio-trace-core/src/metadata.rs]: They're the same type; `tokio-trace::Metadata` is a re-export of `tokio-trace-core::Metadata`. I think the issue is that `tokio-trace` depends on the crates.io version of `tokio-trace-core`, rather than the local version, so the changes I've made to the `-core` docs since the last crates.io release aren't reflected in the `tokio-trace` docs. 

I'll publish a `tokio-trace-core` point release before publishing `tokio-trace` 0.2, so I _think_ this should be resolved. In the meantime, you can patch `tokio-trace`'s dependency on `tokio-trace-core` to see the docs as they _should_ be.

File: tokio-trace/src/lib.rs (hunk 3 above)
Comment (jonhoo) [reply tokio-trace/src/lib.rs]: Oh, sorry, let me rephrase: I think it's great to move things out of the top-level docs the way you have. I meant more that given that you're going to do that, the examples that are left behind should probably be _slightly_ more elaborate than the examples that were there before. I don't think it's too problematic to have more complex concepts in the example here, since the user will know that to get a detailed explanation they should go to the module-level docs.

File: tokio-trace/tokio-trace-core/src/metadata.rs (hunk 29 above)
Comment (jonhoo) [reply tokio-trace/tokio-trace-core/src/metadata.rs]: Ohh, I see. No worries, I just generated docs for `tokio-trace-core` explicitly. I think the `[event]` link is still broken though.

Comment (hawkw) [reply tokio-trace/tokio-trace-core/src/metadata.rs]: I'll double check that one!

**Comment by hawkw** [comment]
@jonhoo in order to prepare for the repo move to `tokio-rs/tracing`, I'd like to merge these changes as-is (since I think they're better than the previous docs) and make some of your suggestions in a follow-up branch after the move. WDYT?

**Comment by davidbarsky** [comment]
Opened https://github.com/tokio-rs/tracing/issues/93 to track applying the rest of @jonhoo's suggestions.
